pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+	/// Our pallet's configuration trait. All our types and constants go in here. If the
+	/// pallet is dependent on specific other pallets, then their configuration traits
+	/// should be added to our implied traits list.
+	///
+	/// `frame_system::Config` should always be included.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + timestamp::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Address which holds the customer funds.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Balances Pallet
+		type NativeCurrency: Currency<Self::AccountId>
+			+ ReservableCurrency<Self::AccountId>
+			+ LockableCurrency<Self::AccountId>;
+
+		type Public: Clone
+			+ PartialEq
+			+ IdentifyAccount<AccountId = Self::AccountId>
+			+ Debug
+			+ parity_scale_codec::Codec
+			+ Ord
+			+ scale_info::TypeInfo;
+
+		/// A matching `Signature` type.
+		type Signature: Verify<Signer = Self::Public>
+			+ Clone
+			+ PartialEq
+			+ Debug
+			+ parity_scale_codec::Codec
+			+ scale_info::TypeInfo;
+
+		/// Governance Origin
+		type GovernanceOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	// Simple declaration of the `Pallet` type. It is placeholder we use to implement traits and
+	// method.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// The extrinsic will be used to start a new reward cycle
+		/// # Parameters
+		/// * `origin`: The donor who wants to start the reward cycle
+		/// * `start_block`: The block from which reward distribution will start
+		/// * `end_block`: The block at which last rewards will be distributed
+		/// * `initial_percentage`: The percentage of rewards that can be claimed at start block
+		/// * `reward_id`: The reward id
+		#[pallet::weight(10_000)]
+		pub fn create_reward_cycle(
+			origin: OriginFor<T>,
+			start_block: T::BlockNumber,
+			end_block: T::BlockNumber,
+			initial_percentage: u32,
+			reward_id: u32,
+		) -> DispatchResult {
+			//check to ensure governance
+			T::GovernanceOrigin::ensure_origin(origin.clone())?;
+
+			//check to ensure no duplicate id gets added
+			ensure!(!<InitializeRewards<T>>::contains_key(reward_id), Error::<T>::DuplicateId);
+
+			//check to ensure start block greater than end block
+			ensure!(start_block < end_block, Error::<T>::InvalidBlocksRange);
+
+			//ensure that difference between start of vesting period - current block is greater
+			//than min difference
+			let difference_between_start_and_current_block = start_block
+				.saturated_into::<u128>()
+				.saturating_sub(<frame_system::Pallet<T>>::block_number().saturated_into::<u128>());
+			ensure!(
+				difference_between_start_and_current_block >
+					MIN_DIFFERENCE_BETWEEN_START_AND_END_BLOCK,
+				Error::<T>::InvalidBlocksRange
+			);
```

Comment:
Remove this check as the crowd loan start block will be an already mined block and not an upcoming block.